controller/bindings/connman/connman.ml

```diff
@@ -86,56 +86,128 @@ struct
 
   type t = input OBus_object.t
 
+  type manager_error =
+      | InvalidKey
+      | ConnectFailed
+      | Blocked
+      (* see manager_error_of_string
+      | OutOfRange
+      | PinMissing
+      | DhcpFailed
+      | LoginFailed
+      | AuthFailed
+      *)
+      | Unknown of string
+
+  type agent_error =
+      | PassMissing
+      | MissingRequestedInputs of string list
+      | BrowserRequested
+      | UnexpectedFailure of exn
+
+  type reported_error =
+      | ManagerError of manager_error
+      | AgentError of agent_error
+
   let request_input input service (fields: (string * OBus_value.V.single) list) =
     let%lwt () = Logs_lwt.debug ~src:log_src
         (fun m -> m "input requested from agent for service %s"
             (String.concat "/" service)
         )
     in
-    match input with
-    | Passphrase p ->
-      (match List.assoc_opt "Passphrase" fields with
-       | Some _ ->
-         return [ "Passphrase", p |> OBus_value.C.(make_single basic_string)]
-       | None ->
-         let%lwt () = Logs_lwt.err ~src:log_src
-             (fun m -> m "Passphrase available as input but not being requested")
-         in
-         OBus_error.make "net.connman.Agent.Error.Canceled" "."
-         |> Lwt.fail
-      )
-    | None ->
-      let%lwt () = Logs_lwt.err ~src:log_src
-          (fun m -> m "input requested from agent but none available.")
-      in
-      OBus_error.make "net.connman.Agent.Error.Canceled" "No input available."
-      |> Lwt.fail
-
+    let pass_requested = List.assoc_opt "Passphrase" fields in
+    match (input, pass_requested) with
+        | (Passphrase p, Some _) ->
+            (* TODO: in theory manager can request passphrase _and_ other
+               fields, so this case is kind of incomplete? *)
+            return @@ Ok [ "Passphrase", p |> OBus_value.C.(make_single basic_string)]
```

Comment:
> We could give up with an error when more than Passphrase is requested, but maybe that's too pessimistic

Looked at [connman's docs](https://git.kernel.org/pub/scm/network/connman/connman.git/tree/doc/agent-api.txt?h=1.42#n47) on `Agent.RequestInput`, and it's actually quite straightforward: connman flags each non-optional requested property as `Requirement: mandatory`, so we can check for those and error right away if there's anything mandatory beyond a passphrase. This is no longer "pessimistic" since connman will outright refuse to connect to if you fail to provide these inputs.  

From the docs: 

> This method gets called when trying to connect to
> a service and some extra input is required. For
> example a passphrase or the name of a hidden network.

Side-note, but If I understand correctly, the hidden network case is unsupported anyway, since it will simply not show up on the AP list.

> and memorize the list of requested inputs, so we can report it in case of subsequent failure

This would be a bit tricky to implement this, since we'd have to thread through a lot of state and it will make the code even more complex. I would just err to the user right away as you suggested.

Reply:
> connman flags each non-optional requested property as `Requirement: mandatory`, so we can check for those and error right away if there's anything mandatory beyond a passphrase. This is no longer "pessimistic" since connman will outright refuse to connect to if you fail to provide these inputs.

OK, sounds good

> If I understand correctly, the hidden network case is unsupported anyway, since it will simply not show up on the AP list.

That's right, we currently don't support them.

> This would be a bit tricky to implement this, since we'd have to thread through a lot of state and it will make the code even more complex. I would just err to the user right away as you suggested.

OK

Reply:
Implemented in 48c66c68fa0